### PR TITLE
Lockout slashing implementation

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

No mechanism to detect and punish validators that vote for concurrent forks.

#### Summary of Changes

1. Compare two VoteStates and decide if there are slashable votes between them.
2. TBD

Fixes #
tag: @carllin

### CODE REVIEW

File: programs/vote/src/vote_state.rs [hunk 1]
```diff
@@ -373,6 +375,67 @@ impl VoteState {
...
+        if test.votes != newest.votes {
+            let i = 0;
+            let j = 0;
+            while i < newest.len() && j < oldest.len() {
+                if newest.votes[i] != oldest.votes[j] {
+                    if newest.votes[i] < oldest.votes[j]  {
+                        slashable.push(newest.votes[i]);
+                        i+=1;
+                        continue;
+                    } else {
```
Comment (carllin) [inline programs/vote/src/vote_state.rs]: I think this is covered, but we don't need special handling of when newest.votes[i] == oldest.votes[j] (leader made two transmissions on different forks). It seems like that case naturally falls out of these ones, but just double checking

File: programs/vote/src/vote_instruction.rs [hunk 2]
```diff
@@ -164,7 +170,7 @@ pub fn withdraw(
 }
 
 pub fn process_instruction(
-    _program_id: &Pubkey,
+    program_id: &Pubkey,
```
Comment (rob-solana) [inline programs/vote/src/vote_instruction.rs]: is this used now?

File: programs/vote/src/vote_state.rs [hunk 3]
```diff
@@ -128,6 +128,7 @@ pub struct VoteState {
 
     /// most recent timestamp submitted with a vote
     pub last_timestamp: BlockTimestamp,
+    pub has_been_slashed: bool,
```
Comment (rob-solana) [inline programs/vote/src/vote_state.rs]: can you put this in epoch_credits instead of in the top level state?

Comment (aeyakovenko) [reply programs/vote/src/vote_state.rs]: @rob-solana Is that a better spot?  I need to scan it and disable it form consensus.

File: programs/vote/src/vote_instruction.rs (hunk 2 above)
Comment (aeyakovenko) [reply programs/vote/src/vote_instruction.rs]: afaik, no.

File: programs/vote/src/vote_state.rs [hunk 4]
```diff
@@ -373,6 +374,70 @@ impl VoteState {
...
+        let not_slashed = if test.votes.last().unwrap() == slashable_slot {
+            //slashable slot is put on top of the stack
+            //the previous vote must be popped
+            self.votes.is_empty()
+                || test.votes[..(test.votes.len() - 1)] == self.votes[..self.votes.len() - 1]
+        } else {
+            //all slashable slots got popped
+            test.votes == self.votes
+        };
+        self.slashed = !not_slashed;
```
Comment (rob-solana) [inline programs/vote/src/vote_state.rs]: to store slashed with the epoch for the slot, you need sysvar::epoch_schedule

File: programs/vote/src/vote_state.rs [hunk 5]
```diff
@@ -489,6 +554,43 @@ pub fn process_vote(
...
+        if tx.instruction_program(i) != program_id {
+            continue;
+        }
+        if let VoteInstruction::Vote(slot) = limited_deserialize(ix.data) {
+            //verify that account matches mine
+            if tx.key(i, 0) != vote_account.key {
+                continue;
+            }
+            //verify that the transaction was signed by the expected signers
+            vote_state::verify_authorized_signer(&vote_state.authorized_voter, &signers)?;
```
Comment (rob-solana) [inline programs/vote/src/vote_state.rs]: I think this means you can escape a slashing if you update your authorized voter key right after trying to cheat?  we might need to limit how often authorized_voter can change so we can keep a list of them

Comment (aeyakovenko) [reply programs/vote/src/vote_state.rs]: yea, key rotation and slashing is a huge Pia.

Comment (aeyakovenko) [reply programs/vote/src/vote_state.rs]: #7555

File: programs/vote/src/vote_state.rs (hunk 4 above)
Comment (aeyakovenko) [reply programs/vote/src/vote_state.rs]: @rob-solana I was thinking once slashed they have to unstake and restake.  the VoteAccount is dead.

File: programs/vote/src/vote_state.rs (hunk 3 above)
Comment (rob-solana) [reply programs/vote/src/vote_state.rs]: is the account disabled from consensus forever?

File: programs/vote/src/vote_state.rs (hunk 4 above)
Comment (rob-solana) [reply programs/vote/src/vote_state.rs]: ok, that's simple.  dead for an epoch.  all unclaimed rewards are gone.

File: programs/vote/src/vote_instruction.rs (hunk 2 above)
Comment (rob-solana) [reply programs/vote/src/vote_instruction.rs]: I don't see it referenced in this function...  does this compile?

File: programs/vote/src/vote_state.rs (hunk 3 above)
Comment (aeyakovenko) [reply programs/vote/src/vote_state.rs]: yea, I think so

File: programs/vote/src/vote_instruction.rs (hunk 2 above)
Comment (aeyakovenko) [reply programs/vote/src/vote_instruction.rs]: not yet

**Pull request closed without merging**